### PR TITLE
[RF] Optimise importing of large expression trees.

### PR DESCRIPTION
**Pull request by hageboeck**
Got a report about very slow importing of objects into workspaces. Turns out that considerable time is wasted with finding objects by name in vectors (which unfortunately cannot be sorted by name).
This adds a `std::unordered_map` for heavy searching and modernises old-style loops in RooWorkspace::import().

[prof_20200526.pdf](https://github.com/root-project/root/files/4683875/prof_20200526.pdf)

### CODE REVIEW

File: roofit/roofitcore/src/RooAbsCollection.cxx [hunk 1]
```diff
@@ -417,10 +419,9 @@ RooAbsArg *RooAbsCollection::addClone(const RooAbsArg& var, Bool_t silent)
 
   // add a pointer to a clone of this variable to our list (we now own it!)
   auto clone2 = static_cast<RooAbsArg*>(var.Clone());
-  if (clone2) _list.push_back(clone2);
-  if (_allRRV && dynamic_cast<const RooRealVar*>(&var)==0) {
-    _allRRV=kFALSE ;
-  }
+  assert(clone2);
```
Comment (eguiraud) [inline roofit/roofitcore/src/RooAbsCollection.cxx:421]: the switch from `if(clone2)` to `assert(clone2)` is a change in behavior, is it intentional?

File: roofit/roofitcore/src/RooAbsCollection.cxx [hunk 2]
```diff
@@ -1281,3 +1292,67 @@ std::unique_ptr<RooAbsCollection::LegacyIterator_t> RooAbsCollection::makeLegacy
...
+  if (item != _nameToItemMap->end()) {
+    // Have an element. Check that it didn't get renamed.
+    if (item->second->namePtr() == item->first) {
+      return item->second;
+    } else {
+      // Item has been renamed / replaced.
+      _nameToItemMap->erase(item);
+      if (auto arg = findUsingNamePointer(_list, namePtr)) {
+        (*_nameToItemMap)[arg->namePtr()] = arg;
+      }
```
Comment (eguiraud) [inline roofit/roofitcore/src/RooAbsCollection.cxx:1352]: so the output of `tryFastFind` can be that `_nameToItemMap` is updated to have the element with `namePtr`, but the function returns nullptr anyway..? i.e. I would expect a `return arg` here

File: roofit/roofitcore/src/RooWorkspace.cxx [hunk 3]
```diff
@@ -647,26 +626,18 @@ Bool_t RooWorkspace::import(const RooAbsArg& inArg,
...
+  for (const auto node : *cloneSet2) {
     if (node->importWorkspaceHook(*this)) {
       coutE(ObjectHandling) << "RooWorkSpace::import(" << GetName() << ") ERROR object named " << node->GetName()
 			        << " has an error in importing in one or more of its auxiliary objects, aborting" << endl ;
       return kTRUE ;
     }
   }
-  iter->Reset() ;
-
-  if (cloneSet2->getSize()+_allOwnedNodes.getSize() > 999) _allOwnedNodes.setHashTableSize(1000);
```
Comment (eguiraud) [inline roofit/roofitcore/src/RooWorkspace.cxx:664]: was the removal of this line intentional? if yes, should it maybe be clarified in the commit message or even be moved to its own commit?

File: roofit/roofitcore/src/RooDirItem.cxx [hunk 4]
```diff
@@ -77,7 +77,7 @@ void RooDirItem::appendToDir(TObject* obj, Bool_t forceMemoryResident)
     if (notInMemNow) {
       gDirectory->cd(pwd) ;    
     }
-
+*/
```
Comment (eguiraud) [inline roofit/roofitcore/src/RooDirItem.cxx:80]: leaving commented code in is typically not considered good practice (also not sure whether it was intentional)

File: roofit/roofitcore/src/RooAbsCollection.cxx (hunk 1 above)
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsCollection.cxx:421]: Yes, the variable should always clone. The `if` should always have been true, and was therefore unnecessary at best. To assert the assumption that a `Clone` in ROOT will not return `nullptr` (not aware that ROOT would do this anywhere), I left an `assert` in place. Makes sense?

File: roofit/roofitcore/src/RooAbsCollection.cxx (hunk 2 above)
Comment (hageboeck) [reply roofit/roofitcore/src/RooAbsCollection.cxx:1352]: Good catch, thanks!
It will still work, but it's now doing a linear search twice.

File: roofit/roofitcore/src/RooDirItem.cxx (hunk 4 above)
Comment (hageboeck) [reply roofit/roofitcore/src/RooDirItem.cxx:80]: Yes, I should take it out.

File: roofit/roofitcore/src/RooWorkspace.cxx (hunk 3 above)
Comment (hageboeck) [reply roofit/roofitcore/src/RooWorkspace.cxx:664]: Intentional, yes. SetHashTableSize wasn't doing anything, but could split the commit.

File: roofit/roofitcore/src/RooAbsCollection.cxx (hunk 1 above)
Comment (eguiraud) [reply roofit/roofitcore/src/RooAbsCollection.cxx:421]: yes, thanks
